### DISCUSSION TITLE
Support for multi-users

### DISCUSSION THREAD

**Discussion by guipace** [vogler/free-games-claimer#74] [Ideas]
It would be great if this application could support more than one user. I would like to be able to add credentials for less tech-savvy friends so their games can be claimed automatically as well. At the moment the workaround is to run multiple containers, but ideally it could be done all in one with this feature.

Thanks for the great work! Much better than having to complete captchas for epic!

**Comment by vogler** [comment]
You can either use a different docker volume for each user or use the same volume but run with different `BROWSER_DIR`, `EMAIL` etc.
I assume most people just want to claim one account for each store, so adding logic for managing several users in one instance is likely complicating things too much for the majority. Maybe once there's a web interface for easy configuration.

**Reply by twistedsanity** [reply]
Sorry for my ignorance, I couldn't quite figure the correctr usage of the browser_dir for multiple epic accounts so currently running multiple containers. I assume there is a more optimum way around this?
Even in this way it is working so much better than my previous solution, thank you.

**Reply by vogler** [reply]
Here's an example:
```console
$ docker run --rm -it -p 6080:6080 -v fgc:/fgc/data --pull=always -e BROWSER_DIR=data/browser1 -e EMAIL=email1 ghcr.io/vogler/free-games-claimer node epic-games
$ docker run --rm -it -p 6080:6080 -v fgc:/fgc/data --pull=always -e BROWSER_DIR=data/browser2 -e EMAIL=email2 ghcr.io/vogler/free-games-claimer node epic-games
```
This will create two different browser profiles inside the `fgc` volume which you can use to login with different accounts. The default path is `data/browser`.
Terminology: a container is ephemeral, i.e., created for any run, but removed right after due to the `--rm` flag. Since data is stored in the `fgc` volume and not in the container, they are stateless/throwaway.

**Comment by DevXen** [comment]
I'm not sure I can do that browser_dir on unraid?

**Reply by vogler** [reply]
Why not? Just configure a run with separate `BROWSER_DIR` for each user.

**Comment by KairuByte** [comment]
The main reason I found my way to this discussion, was because I was looking into wrapping this into a Home Assistant addon. However, you don't have the ability within Home Assistant to dynamically change settings once a container has been created. This means you can neither change `BROWSER_DIR` nor the volume.

I suppose this is just me adding a +1 to multiple user support, even if the chances of it happening are small.

**Reply by vogler** [reply]
Can't you run one container per user? The container is removed afterwards anyway.

Multiple users are already supported, just with their own config. What solution do you have in mind? If you want to define several users in one config, you'd need to change it to something like JSON instead of using dotenv which would make configuration via environment variables complicated.

**Reply by KairuByte** [reply]
I am a novice when it comes to HA, so I may be getting things slightly wrong. But it essentially loads the configuration defined into portainer, and spins up that container when the add-on is started. You are also limited to one "install" of the add-on, unless you make a new add-on per user. This means you can technically have multiple containers, but would require an arbitrary number of add-on listings since some people may only want to run one, and some may want to run 5.

You are correct that you are unable to change the environment variables in this case as well. Which is actually something I was going to bring up during a feature request I had not fully thought through just yet. I was actually going to suggest allowing config load through *either* environment variables, *or* config file, with the environment variables taking precedent if both are found. This would (ideally) allow full configuration from within the volume, and would have the added benefit of being able to list out all config options if so desired.

**Reply by vogler** [reply]
> allowing config load through _either_ environment variables, _or_ config file, with the environment variables taking precedent if both are found

That's already what's happening. You can have one volume per user, each with their own `data/config.env` and `data/browser`.

**Reply by KairuByte** [reply]
The issue is that you cant specify the volume beyond initial setup. So you can't dynamically swap between volumes between runs.
